File: Chapter09/Challenge05/summing_binaries.js
const p = 17;
const q = 20;

const bitsSum = (p, q) => {
	let and = p & q;
	let xor = p ^ q;

	while (and) {
		and = and << 1;
		
		const tmp = xor ^ and;
		and = xor & and;
		xor = tmp;
	}

	return `${xor.toString(2)}(${xor})`;
};

console.log(`\t${p.toString(2)}(${p})
+\t${q.toString(2)}(${q})
------------------
\t${bitsSum(p, q)}`);

File: Chapter09/Challenge05/summing_binaries.py
def sum(p: int, q: int):
    _and = (p & q)
    xor = p ^ q

    while _and != 0:
        _and = _and << 1

        t = xor ^ _and
        _and = _and & xor

        xor = t

    return xor


print(sum(2, 2))